docs/enhancements/image-handling-service.md

```diff
@@ -0,0 +1,182 @@
+
+### Goals
+
+* Allow image serving to be scaled independently of the rest of the application.
+* Remove performance impact due to generating and serving images from the assisted service.
+* Make minimal changes to existing (or proposed) APIs.
+
+### Non-Goals
+
+* Serve a discovery image directly to a BMC
+
+## Proposal
+
+The new service (image service) will handle customizing and serving the customized
+image to users. It will query assisted service to retrieve the information that needs
+to be added to the image based on the information passed in the request by the user.
+
+The assisted service will retain the `POST /clusters/{cluster_id}/downloads/image`
+endpoint which will save the image information to its database. In the case
+of the kube API, the controllers will continue to save image information from the
+various CRs as they do currently.
+
+The image service will expose a single API endpoint to download an image. Initially
+this should be compatible with the existing `GET /clusters/{cluster_id}/downloads/image`
+assisted service endpoint. When a user makes a request to this endpoint the
+image service will fetch the ignition and ramdisk (if necessary) from the
+assisted service and stream the customized iso to the client using the
+appropriate template image as a base. This will involve adding an endpoint to
+assisted service to expose the ramdisk data.
+
+### User Stories
+
+#### Story 1
+
+As a user deploying and running assisted service, I need image serving to be
+horizontally scalable and to not impact performance of the assisted service API.
+
+#### Story 2
+
+An additional deployment and service will be managed for all deployment methods.
+This includes SaaS, operator, and CI (minikube). Whether an additional image is
+created is still an open question.
+
+### Implementation Details/Notes/Constraints
+
+#### Request Flow
+
+1. Image service receives a request to download an image
+  - This request contains, as query parameters, the assisted service auth token, the image type (minimal/full), and image version
+2. Image service queries assisted service for the image ignition
+3. Assisted service generates the ignition and serves it
+4. If the iso is minimal, the image service queries the assisted service for the initrd
+5. Assisted service generates the initrd and serves it
+6. Image service streams base iso with ignition and initrd embedded to the user
+
+#### Authentication and Authorization
+
+The image service will not need to implement authentication or authorization
+directly as it doesn't manage any user data. The user will provide a token to
+the image service and the image service will directly pass that token to
+assisted service on each call. Assisted service will validate the token before
+giving the image service any user information.
+
+#### Template Images
+
+Currently, when the assisted service starts, it downloads and caches each RHCOS
+image specified in the `OPENSHIFT_VERSIONS` environment variable. It also
```

Comment:
Is there any automation around invalidating this cache, e.g if the upstream version changes, or is the expectation that some human intervention is required to bounce the service?

Reply:
> the expectation that some human intervention is required to restart the service

This. Right now a mapping from OCP version to rhcos url is provided to assisted service as an env var.
ref: https://github.com/openshift/assisted-service/blob/master/data/default_ocp_versions.json